crates/common/tedge_utils/src/fs.rs

```diff
@@ -8,41 +8,97 @@ use std::path::PathBuf;
+}
+
+impl<T, E: Into<std::io::Error>> ErrContext<T> for Result<T, E> {
+    fn context(self, context: String, file: impl AsRef<Path>) -> Result<T, AtomFileError> {
+        self.map_err(|err| AtomFileError::WriteError {
+            file: Box::from(file.as_ref()),
+            context,
+            source: err.into(),
+        })
+    }
+}
+
+pub struct TempFile(PathBuf);
+
+impl Drop for TempFile {
+    fn drop(&mut self) {
+        let _ = std_fs::remove_file(&self.0);
+    }
+}
+
 /// Write file to filesystem atomically using std::fs synchronously.
 pub fn atomically_write_file_sync(
     dest: impl AsRef<Path>,
     mut reader: impl Read,
-) -> std::io::Result<()> {
+) -> Result<(), AtomFileError> {
     let dest_dir = parent_dir(dest.as_ref());
     // FIXME: `.with_extension` replaces file extension, so if we used this
     // function to write files `file.txt`, `file.bin`, `file.jpg`, etc.
     // concurrently, then this will result in an error
-    let tempfile = PathBuf::from(dest.as_ref()).with_extension("tmp");
+    let tempfile = TempFile(PathBuf::from(dest.as_ref()).with_extension("tmp"));
 
     // Write the content on a temp file
     let mut file = std_fs::OpenOptions::new()
         .write(true)
         .create_new(true)
-        .open(&tempfile)?;
-
-    if let Err(err) = std::io::copy(&mut reader, &mut file) {
-        let _ = std_fs::remove_file(tempfile);
-        return Err(err);
-    }
+        .open(&tempfile.0)
+        .context(
+            format!("could not create the temporary file {:?}", tempfile.0),
+            &dest,
+        )?;
+
+    std::io::copy(&mut reader, &mut file).context(
+        format!(
+            "could not copy the content to the temporary file {:?}",
+            tempfile.0,
+        ),
+        &dest,
+    )?;
 
     // Ensure the content reach the disk
-    file.flush()?;
-    file.sync_all()?;
+    file.flush().context(
+        format!(
+            "could not flush the content of the temporary file {:?}",
+            tempfile.0,
+        ),
+        &dest,
+    )?;
+
+    file.sync_all().context(
+        format!("could not save the temporary file {:?} to disk", tempfile.0,),
+        &dest,
+    )?;
 
     // Move the temp file to its destination
-    if let Err(err) = std_fs::rename(&tempfile, dest) {
-        let _ = std_fs::remove_file(tempfile);
-        return Err(err);
-    }
+    std_fs::rename(&tempfile.0, &dest).context(
+        format!(
+            "could not move the file from {:?} to {:?}",
+            tempfile.0,
+            dest.as_ref(),
+        ),
+        &dest,
+    )?;
 
     // Ensure the new name reach the disk
-    let dir = std::fs::File::open(dest_dir)?;
-    dir.sync_all()?;
+    let dir =
+        std::fs::File::open(dest_dir).context("could not open the directory".to_string(), &dest)?;
+
+    dir.sync_all()
+        .context("could not save the file to disk".to_string(), &dest)?;
```

Comment:
issue: because argument to `.context()` is a `String`, and not `impl Fn() -> String`, we're always allocating strings for all possible errors, even if they don't happen!

We should instead use a closure, so we're only allocating context when it's necessary:

```rust
pub trait ErrContext<T> {
    fn with_context(
        self,
        context: impl Fn() -> String,
        file: impl AsRef<Path>,
    ) -> Result<T, AtomFileError>;
}

impl<T, E: Into<std::io::Error>> ErrContext<T> for Result<T, E> {
    fn with_context(
        self,
        context: impl Fn() -> String,
        file: impl AsRef<Path>,
    ) -> Result<T, AtomFileError> {
        self.map_err(|err| AtomFileError::WriteError {
            file: Box::from(file.as_ref()),
            context: context(),
            source: err.into(),
        })
    }
}
```

In anyhow, `.context()` can take in a `&str`, which is statically allocated, but for `String` you should use `.with_context()`, which similarly takes in a closure. If we give up formatting the context string, and only use a string slice, we could use `&'static str` instead:

```rust
#[derive(Debug, thiserror::Error)]
pub enum AtomFileError {
    #[error("Writing the content to the file {file:?} failed: {context:?}. source={source:?}")]
    WriteError {
        file: Box<Path>,
        context: &'static str,
        source: std::io::Error,
    },
}

pub trait ErrContext<T> {
    fn context(
        self,
        context: &'static str,
        file: impl AsRef<Path>,
    ) -> Result<T, AtomFileError>;
}

impl<T, E: Into<std::io::Error>> ErrContext<T> for Result<T, E> {
    fn context(
        self,
        context: &'static str,
        file: impl AsRef<Path>,
    ) -> Result<T, AtomFileError> {
        self.map_err(|err| AtomFileError::WriteError {
            file: Box::from(file.as_ref()),
            context,
            source: err.into(),
        })
    }
}
```